src/Pact/Types/Command.hs

```diff
@@ -253,37 +254,48 @@ instance FromJSON UserSig where
   {-# INLINE parseJSON #-}
 
 
+data CommandResult = CommandResult {
+  _crReqKey :: RequestKey,
+  _crTxId :: Maybe TxId,
+  _crResult :: Value
+  } deriving (Eq,Show)
 
+-- | Actual value of a `CommandResult`.
+data CommandValue a
+  = CommandSuccess a
+  | CommandFailure CommandError
+  deriving (Eq, Show)
 
-
-
+-- | Error details for `CommandValue`.
 data CommandError = CommandError {
       _ceMsg :: String
     , _ceDetail :: Maybe String
-}
-instance ToJSON CommandError where
-    toJSON (CommandError m d) =
-        object $ [ "status" .= ("failure" :: String)
-                 , "error" .= m ] ++
-        maybe [] ((:[]) . ("detail" .=)) d
-
-newtype CommandSuccess a = CommandSuccess { _csData :: a }
-  deriving (Eq, Show)
-
-instance (ToJSON a) => ToJSON (CommandSuccess a) where
-    toJSON (CommandSuccess a) =
-        object [ "status" .= ("success" :: String)
-               , "data" .= a ]
-
-instance (FromJSON a) => FromJSON (CommandSuccess a) where
-    parseJSON = withObject "CommandSuccess" $ \o ->
+  } deriving (Eq, Show)
+
+
+instance ToJSON a => ToJSON (CommandValue a) where
+    toJSON = \case
+      CommandSuccess a -> object
+        [ "status" .= ("success" :: Text)
+        , "data"   .= a
+        ]
+      CommandFailure (CommandError {..}) -> object $
+        [ "status" .= ("failure" :: Text)
+        , "error" .= _ceMsg
+        ]
+        ++ maybe [] (pure . ("detail" .=)) _ceDetail
+
+instance FromJSON a => FromJSON (CommandValue a) where
+  parseJSON = withObject "CommandValue" $ \o -> do
+    status <- o .: "status"
+    case status of
+      "failure" -> do
+        msg <- o .: "error"
+        detail <- o .:? "detail"
+        pure $ CommandFailure $ CommandError msg detail
+      "success" -> do
```

Comment:
I can see what @slpopejoy is saying here re: just using `Either CommandError a` here. They're roughly equivalent, but I like the idea of just casing on `Left | Right` as opposed to using Yet Another Record to handle it. Just a newtype over `Either` gives us enough imo. If at a later date we need to change it, that's fine.

Reply:
If we go with `Either`, I think we should accept the changed wire format. I don't see how a `newtype` wrapper would be better than a standalone data type in this case. Also, what would be its name? `CommandValue` again? In this case all we got is that we now need a two level unwrapping for accessing values.

I wasn't happy with the `CommandValue` either: So as it seems nobody has a problem with the changed wire format, I will replace it with a plain `Either` then :-)